### PR TITLE
Add puma plugin

### PR DESCRIPTION
**Pull request by npezza93**
This adds a puma plugin that runs watch mode only in development.

### CODE REVIEW

File: lib/puma/plugin/tailwindcss.rb [hunk 1]
```diff
@@ -0,0 +1,65 @@
+require "puma/plugin"
+
+Puma::Plugin.create do
+  attr_reader :puma_pid, :tailwind_pid, :log_writer
+
+  def start(launcher)
+    return unless Rails.env.development?
```
Comment (brunoprietog) [inline lib/puma/plugin/tailwindcss.rb:7]: Wouldn't this condition be better in the Puma configuration? Only include the plugin if the environment is a development environment.

File: lib/puma/plugin/tailwindcss.rb [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+    @log_writer = launcher.log_writer
+    @puma_pid = $$
+    @tailwind_pid = fork do
+      Thread.new { monitor_puma }
+      system(*Tailwindcss::Commands.watch_command)
+    end
+
+    launcher.events.on_stopped { stop_solid_queue }
```
Comment (brunoprietog) [inline lib/puma/plugin/tailwindcss.rb:16]: Ah, taken from Solid Queue. This needs to be changed

File: lib/puma/plugin/tailwindcss.rb [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+    in_background do
+      monitor_tailwind
+    end
+  end
+
+  private
+    def stop_solid_queue
```
Comment (brunoprietog) [inline lib/puma/plugin/tailwindcss.rb:24]: Same here

**Review by brunoprietog — COMMENTED** [review]
This is great! Goodbye `bin/dev`. It would be nice to change the install generator to default to include the plugin in the Puma configuration in the development environment.

File: lib/puma/plugin/tailwindcss.rb (hunk 2 above)
Comment (npezza93) [reply lib/puma/plugin/tailwindcss.rb]: Whoops 😅

File: lib/puma/plugin/tailwindcss.rb (hunk 1 above)
Comment (npezza93) [reply lib/puma/plugin/tailwindcss.rb]: Yea i went back and forth. I choose this one so people didnt have to worry about it in the puma config but i think having it be explicit in the puma config gives some flexibility. Changed it

**Comment by npezza93** [comment]
@brunoprietog  Should have everything addressed! I also removed the bin/dev and procfile stuff. Let me know if you want to keep that around

File: lib/install/tailwindcss.rb [hunk 4]
```diff
@@ -38,19 +38,8 @@
...
-
-  say "Ensure foreman is installed"
-  run "gem install foreman"
-end
-
-say "Add bin/dev to start foreman"
-copy_file "#{__dir__}/dev", "bin/dev"
-chmod "bin/dev", 0755, verbose: false
+say "Add puma plugin"
+append_to_file "config/puma.rb", %(plugin :tailwindcss if Rails.env.development?)
```
Comment (brunoprietog) [inline lib/install/tailwindcss.rb:42]: Kind of nitpicky, but what do you think about adding a blank line before and after the plugin? And a brief comment saying what the plugin does. For example, Rails default has this plugin like this:

```ruby
# Allow puma to be restarted by `bin/rails restart` command.
plugin :tmp_restart
```

**Review by brunoprietog — COMMENTED** [review]
I'm not a maintainer, I just leave my comments that I think could be useful to help this merge more quickly.

I would really love to leave bin/dev, and actuallty this same change could be very good for dartsas-rails, jsbundling-rails and cssbundling-rails if the maintainers agree.

I think it would be good to add a description to the PR. Leave bin/dev is a very convincing argument at least for me.

**Comment by brunoprietog** [comment]
Nice! I have no further comments from me at least. Hopefully it will merge soon and we will have the same for the other gems that force us to use bin/dev.

**Comment by flavorjones** [comment]
I think I understand why adding Puma support is valuable; but what's the rationale for deleting `lib/install/Procfile.dev` and `lib/install/dev`? That will break the developer experience for people using foreman today.

**Comment by brunoprietog** [comment]
It won't break it, it would only be for new applications. `bin/dev` is horrible for debugging, and I personally don't like using foreman because it adds extra information that I don't need in the terminal. You should only need to run `bin/rails server` and that's it.

**Comment by flavorjones** [comment]
There is a developer workflow that is well-documented and expected to work. Unless the files prevent users from using a puma plugin, they should not be removed. And if they do, I want to understand why.

**Comment by brunoprietog** [comment]
How could you debug using `bin/dev`? Whenever I need to debug, I have to use `rails server` separately. That for me at least is quite annoying.

In my opinion, everything should just work in a single terminal. Why would you want to have two ways if one works better than the other? I don't understand why it would be better to use `bin/dev`.

I mean, why would you want to use foreman? That means adding one more dependency. It also means that in some cases, newvie developers have to use `bin/dev` and in others `rails server`. Explaining this to a newbie coming from node.js for example is not so easy, being that they can run everything in a single process there.

**Comment by npezza93** [comment]
@flavorjones How would you feel about reverting all but the puma plugin and have the discussion around bin/dev in a separate PR?

**Comment by flavorjones** [comment]
@npezza93 Yup, that would be great, thank you!

**Comment by npezza93** [comment]
@flavorjones   Should be all set!

**Comment by flavorjones** [comment]
@npezza93 Thank you! I've squashed your commits to leave a clean history, and I've reworked the README a bit to feature the Puma plugin more prominently.

This is great work, thank you! Worked great for me when I tried it out on my local dev environment.

**Comment by airblade** [comment]
> I mean, why would you want to use foreman?

@brunoprietog So you can spin up other processes too, such as Redis or Sidekiq.

**Comment by brunoprietog** [comment]
Redis can be started as a service just like PostgreSQL or MySQL. And Sidekiq you can also manage it with puma, so a single command is enough and you can easily debug.

**Comment by hdittmar** [comment]
I tried adding the `plugin :tailwindcss...` configuration line mentioned in the Readme into my puma.rb file and since then, puma crashes on launch, for more info see here: https://github.com/rails/tailwindcss-rails/issues/319

**Pull request merged**